prusti-tests/tests/verify/fail/predicates/doesnt-hold.rs

```diff
@@ -0,0 +1,11 @@
+use prusti_contracts::*;
+
+#[predicate]
+fn false_p() -> bool {
+    false
+}
+
+#[ensures(false_p())] //~ ERROR precondition of pure function call might not hold
```

Comment:
It's good that this raises an error and with a reasonable position, but now that I think about it, the error message is actually a bit confusing. "precondition of pure function" here refers to `false_p`, which is specified without a precondition as far as the user is concerned. Changing this would actually be a bit more involved:

- during macro processing of `#[predicate]`, mark the resulting function with some new prusti-reserved attribute
- after typing, predicates should be identified by the `SpecCollector`
  - `ProcedureSpecifications` should get a new field `predicate: bool` if the attribute was found in the HIR
- `ErrorCtxt` needs a new variant for predicate calls
- `encode_specified_pure_function_call` (in `procedure_encoder`) needs to identify when a pure function call is a call to a predicate (which will now be stored in the specification set)

We can wait until Monday if this is unclear, it also makes me a bit uneasy to introduce a relatively large set of changes to fix one error message…

Reply:
yep, i think i understand the problems you see here, but i agree with the "large set of changes to fix one error message". i'd rather get this feature landed at all first, and do more involved stuff like this subsequently.